### ISSUE TITLE
transmission: package should not set permissions

### ISSUE THREAD

**Issue by monokrome** [NixOS/nixpkgs#147239]
### Describe the bug

If a network share is being used for Transmission, all `nixos-rebuild` operations fail because transmission will try to set permissions but network shares (and likely other filesystems) may or may not allow such actions.

### Steps To Reproduce

1. Set up a network share without chmod permissions, for instance NFS on a Synology NAS (which doesn't allow the chmod)
2. Add transmission, using the aforementioned mount for it's file
3. Run `nixos-rebuild switch --upgrade`
4. Watch `nixos-rebuild` fail due to this issue

### Expected behavior

Transmission expects the sysadmin to set up their permissons with custom paths, or - if that is not idea - provides an option to prevent permission setting.

### Screenshots

<img width="786" alt="Screen Shot 2021-11-24 at 00 38 41" src="https://user-images.githubusercontent.com/242997/143203464-0f4bda78-f056-4f4d-a646-a4d322ee04a2.png">

### Notify maintainers

@ju1m 

### Metadata

```
these paths will be fetched (0.05 MiB download, 0.28 MiB unpacked):
  /nix/store/1bpw671p7r93sp3sbdxq9ly6h4npx5ff-bash-interactive-4.4-p23-dev
copying path '/nix/store/1bpw671p7r93sp3sbdxq9ly6h4npx5ff-bash-interactive-4.4-p23-dev' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.79, NixOS, 21.05.4349.09650059d7f (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.16`
 - channels(root): `"nixos-21.05.4349.09650059d7f"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:

```yaml
attribute:
  - "download-dir"
  - "incomplete-dir"
  - "watch-dir"

module: transmission
```

**Comment by aanderse** [comment]
@monokrome thank you very much for providing this use case. Much like we (usually) conditionally provision system users and groups for NixOS modules, we need to conditionally provision data and state directories so that workflows like the one you mentioned here aren't excluded.

@ju1m my suggestion is that we only run `system.activationScripts.transmission-daemon` conditional on `cfg.home == "/var/lib/transmission"`, adding something like the following to the `description` of `home`:

> If left as the default value this directory will automatically be created on system activation, otherwise the sysadmin is responsible for ensuring the directory exists with appropriate ownership and permissions.

@ju1m - what do you think?
@monokrome - would that work for you?

**Comment by monokrome** [comment]
I think that is the ideal use-case in my opinion :) Thanks for the quick reply in the thread 🙏

**Comment by aanderse** [comment]
Can you please paste your `services.transmission` configuration?

**Comment by ju1m** [comment]
That kind of problem with `downloadDirPermissions` [happened before](https://github.com/NixOS/nixpkgs/issues/100074#issuecomment-715659908).

I don't like this option which is deviating from the usual `StateDirectory=/StateDirectoryMode=` settings, I'm in favor of removing this option or at least make it nullable to set `system.activationScripts.transmission-daemon` if and only if `downloadDirPermissions != null`.

By default `/var/lib/transmission` is used, it is created because it's the `home` of the default `transmission.user` but I guess we should also add a `StateDirectory=["transmission"]` and `StateDirectoryMode="750"` which have some logic to change ownership/permissions:

> If the specified directories already exist and their owning user or group do not match the configured ones, all files and directories below the specified directories as well as the directories themselves will have their file ownership recursively changed to match what is configured. As an optimization, if the specified directories are already owned by the right user and group, files and directories below of them are left as-is, even if they do not match what is requested. The innermost specified directories will have their access mode adjusted to the what is specified in RuntimeDirectoryMode=, StateDirectoryMode=, CacheDirectoryMode=, LogsDirectoryMode= and ConfigurationDirectoryMode=
Source : https://www.freedesktop.org/software/systemd/man/systemd.exec.html

**Comment by ju1m** [comment]
@monokrome, @aanderse, I've proposed https://github.com/NixOS/nixpkgs/pull/147324 to begin a fix, please let me know if it fits your expectations or what you think could be done better.

**Comment by monokrome** [comment]
@aanderse ```nix
    transmission = {
      enable = true;

      settings = {
        rpc-bind-address = "127.0.0.1";
        rpc-port = 9091;
        rpc-whitelist = "127.0.0.1";

        download-dir = "/mnt/storage/Multimedia/.t/done";

        incomplete-dir = "/mnt/storage/Multimedia/.t/working";
        incomplete-dir-enabled = true;

        watch-dir = "/mnt/storage/Downloads/Torrents";
        watch-dir-enabled = true;
      };
    };
```

@ju1m I didn't catch all of that, but am I right that the proposed fix would simply stop this from setting permissions unless verbosely asked for?

**Comment by ju1m** [comment]
@monokrome, yes, with https://github.com/NixOS/nixpkgs/pull/147324 `downloadDirPermissions` would default to `null`, and `system.activationScripts.transmission-daemon` would only be enabled if `downloadDirPermissions` is set to some permissions.

**Comment by monokrome** [comment]
I think this is fixed by https://github.com/NixOS/nixpkgs/pull/147324, yeah?

**Comment by bjornfor** [comment]
> I think this is fixed by https://github.com/NixOS/nixpkgs/pull/147324, yeah?

Ok, closing. (If it isn't done, then just add a comment and we'll re-open.)

**Comment by monokrome** [comment]
Thanks @bjornfor 😺

**Comment by wyndon** [comment]
One problem is that `system.activationScripts.transmission-daemon` will not work inside a nixos container. I wonder if there's a way around that.

**Comment by monokrome** [comment]
@wyndon Can you give more details? Why wouldn't it work?

**Comment by wyndon** [comment]
> @wyndon Can you give more details? Why wouldn't it work?

In my experience, `system.activationScripts.*` doesn't do anything within a container. So basically, setting "downloadDirPermissions"  does nothing on my end.

**Comment by aanderse** [comment]
> In my experience, `system.activationScripts.*` doesn't do anything within a container.

Really? 🤔
